#458736 [C#テキスト]6章演習問題

kurosu-being opened the pull request:
第6章 配列とList<T>の操作

主にLINQを使ったコレクションの比較や集計、抽出について学習しました。

kurosu-being committed:
#458736 [C#テキスト]6章演習問題

Practice6-1-1のフォルダを作成しました。

kurosu-being committed:
#458736 [C#テキスト]6章演習問題

Practice6-1-1のフォルダを作成しなおしました。

kurosu-being committed:
#458736 [C#テキスト]6章演習問題

Practice6-1-1のフォルダを作成しなおしました。

kurosu-being committed:
#458736 [C#テキスト]6章演習問題

Program.csの名前をProgram6-1-1に変更しました。

kurosu-being committed:
#458736 [C#テキスト]6章演習問題

Practice6-1-1~Practice6-1-3まで実装しました。

kurosu-being committed:
#458736 [C#テキスト]6章演習問題

Practice6-2-2~Practice6-2-4まで実装しました。

kurosu-being committed:
#458736 [C#テキスト]6章演習問題

Practice6-2-5~Practice6-2-6まで実装しました。

kurosu-being committed:
#458736 [C#テキスト]6章演習問題

Practice6-2-7を実装しました。

kurosu-being committed:
#458736 [C#テキスト]6章演習問題

Practice6-2のプロパティを読み取り専用にしました。
Practice6-2-5において変数名を変更しました。（複数形→単数形）
Practice6-2-6において変数名を変更しました。
Practice6-2-7において変数名を変更しました。

kurosu-being committed:
#458736 [C#テキスト]6章演習問題

Practice6-1-1~Practice6-2-2まで不要な変数を削除しました。

kurosu-being committed:
#458736 [C#テキスト]6章演習問題

不要な変数の消去を行いました。
コードが見やすいように空行を入れました。

kurosu-being committed:
#458736 [C#テキスト]6章演習問題

不要なファイルを削除しました。

kurosu-being committed:
#458736 [C#テキスト]6章演習問題

コンストラクタの引数の説明を追加しました。

kurosu-being committed:
#458736 [C#テキスト]6章演習問題

価格のプロパティをクラス内で変更できるように変更しました。

kurosu-being committed:
#458736 [C#テキスト]6章演習問題

Practice6-2-1以降のBookクラスにおいて、Bookクラスのコメントを追加しました。

k-miyahara-being commented on 章末問題6/Practice6-1-1/Program6-1-1.cs:
```diff
@@ -0,0 +1,17 @@
+            //ビーイングコーディング規約に合わせ問題分のnumbers→wNumbersに変更
+            var wNumbers = new int[] { 5, 10, 17, 9, 3, 21, 10, 40, 21, 3, 35, };
+
+            Console.WriteLine(wNumbers.Max());
```
Max()を使うときに注意してほしいのは、コレクションに要素があるかどうかです。
仮にwNumbersになにも要素がない状態で実行してみると、例外が発生してしまいます。

[追加問題]
要素があるかどうかを判定する処理を追加してみてください。
要素がない場合は、「要素がありません」を出力するようにしてください。

LINQなど既存のメソッド等を使うときは、公式ドキュメントを読んで、どんな例外が発生する可能性があるか確認してみるとよいです。
該当のメソッドにカーソルを当て、F1キーで公式ドキュメントを開くことができます。

kurosu-being replied:
ご指摘ありがとうございます。
追加問題を実装し、要素がない場合に例外が発生しないよう処理を追加しました。
公式ドキュメントを読む際、例外の箇所を読むことがなかったので、今後活用していこうと思います💪

k-miyahara-being commented on 章末問題6/Practice6-1-2/Program6-1-2.cs:
```diff
@@ -0,0 +1,22 @@
+                }
+            } else {
+                Console.WriteLine("コレクション数が1以下です");
+            }
```
コレクションの要素数で事前に判定しているのはとても良いですね！
ここでは、以前お話したガード節を使ってみてはどうでしょう。
要素が2つ未満なら、早期にreturnしてメソッドを抜けてしまったほうが、可読性は上がります。
基本的にコードは上から下へ向かって読んでいくので、条件分岐の際は、条件を考えつつ中の処理を読む必要があるため考えることが多くなります。
ガード節があることで、必要のない条件は思考から省きながら読み進めることができます。

ガード節については → https://note.com/merucari00/n/nef260fe78abf

kurosu-being replied:
ご指摘ありがとうございます！
要素が2つ未満の時、早期にreturnしてメソッドを抜けるように変更いたしました。
実際に実装してみて、本当にやりたい処理や知りたい処理が見やすくなるのが分かりました。
また、参考URLで使用されていたcontinueキーワードが初見でしたので、こちらも適宜使えるようにします。

k-miyahara-being commented on 章末問題6/Practice6-1-3/Program6-1-3.cs:
```diff
@@ -0,0 +1,15 @@
+
+            foreach (string wStringNumber in wNumbers.Select(x => x.ToString()).ToArray()) {
+                Console.WriteLine(wStringNumber);
+            }
```
ToArray()はしなくて良いかなと思います。
以降のコードで配列やリストとして管理する必要があるなら使用すべきですが、今回の場合は使用しなくても問題ありません。
ToList()やToList()は、コレクションIEnumerable<T>を実体化させるメソッドです。
実体化させるということはメモリが消費されるということです。
コレクションの規模が大きいほどメモリ消費によるパフォーマンス低下が懸念されるため、必要ない場合はToList()、ToArray()は使用しないほうが良いかなと思います。

参考 → https://atmarkit.itmedia.co.jp/ait/articles/1409/24/news105.html

kurosu-being replied:
ご指摘ありがとうございます！
ToArray()を削除いたしました。
「コレクションIEnumerableを実体化させる」という観点がなかったので勉強になりました🙇‍♀️

k-miyahara-being commented on 章末問題6/Practice6-1-4/Program6-1-4.cs:
```diff
@@ -0,0 +1,16 @@
+            var wNumbers = new int[] { 5, 10, 17, 9, 3, 21, 10, 40, 21, 3, 35 };
+
+            Console.WriteLine("小さい順に並べた最初の3つの要素:");
+            foreach (int wSortThreeNumber in wNumbers.OrderBy(x => x).Take(3)) {
+                Console.WriteLine(wSortThreeNumber);
+            }
```
wNumbersがnullのとき、例外が発生します。※無理やりwNumbersにnullをセットした場合
LINQを使用するときは、nullになる可能性がないかを確認することが大切です。

メソッドチェーンをしていると、どこかのタイミングで既定値nullが返されて、その結果例外が起こるということがよくあります。
nullになりうる可能性がある場合は、nullチェックとnullになったときどうするかを忘れず実装するように気を付けてみてください。

今回は、8行目で確実にnewしているので11行目時点でwNumbersがnullになることはないですが、
仮にnullがセットされた場合を想定して実装し直してみてください。

kurosu-being replied:
ご指摘ありがとうございます！
コレクションの中身にnullがセットされた時、メッセージで返すように実装いたしました。
LINQを使用する時、メソッドチェーンをしている時は特に気をつける必要があるのですね。気をつけていこうと思います。

k-miyahara-being reviewed (changes requested):
一旦問1まで

kurosu-being committed:
#458736 [C#テキスト]6章演習問題

Practice6-1-1において、追加問題を実装しました。
Practice6-1-2において、ガード節で書き直しました。
Practice6-1-3において、ToArrayメソッドを削除しました。
Practice6-1-4において、コレクションの中身がnullの時エラーメッセージが出るように修正しました。
Practice6-1-5において、コレクションの中身がnullの時エラーメッセージが出るように修正しました。
Practice6-2-1において、本のタイトルが重複していた33行目C#プログラミングの常識→楽しいプログラミング教室に修正しました。また、これ以降のcsにおいても同様に変更しました。
Practice6-2-2において、コレクションの中身がnullの時エラーメッセージが出るように修正しました。
Practice6-2-3において、コレクションの中身がnullの時エラーメッセージが出るように修正しました。
Practice6-2-5において、コメントを一部修正しました。
Practice6-2-6において、コレクションの中身がnullの時エラーメッセージが出るように修正しました。
Practice6-2-7において、コレクションの中身がnullの時エラーメッセージが出るように修正しました。

kurosu-being committed:
#458736 [C#テキスト]6章演習問題

Practice6-1-4において、フォーマットのズレを修正しました。

kurosu-being committed:
#458736 [C#テキスト]6章演習問題

Practice6-2-4以降において、コレクションの中身がnullの時のメッセージを変更しました。（本が登録されていません。→コレクションの中身がnullです。）

k-miyahara-being commented on 章末問題6/Practice6-2-1/Program6-2-1.cs:
```diff
@@ -0,0 +1,43 @@
+            var wTargetBook = wBooks.FirstOrDefault(x => x.Title == "ワンダフル・C#ライフ");
+            //価格はカンマ付き数字に変換しています
+            string wBookInfomation = wTargetBook != null
+                ? $"書籍「{wTargetBook.Title}」の価格は {wTargetBook.Price.ToString("#,0")}円でページ数は{wTargetBook.Pages}Pです" : "指定したタイトルの本が見つかりませんでした。";
```
カンマ付き、nullチェック、ともに良いですね！👍

k-miyahara-being commented on 章末問題6/Practice6-2-2/Program6-2-2.cs:
```diff
@@ -0,0 +1,27 @@
+                Console.WriteLine("コレクションの中身がnullです。");
+                return;
+            }
+            Console.WriteLine($"タイトルに\"C#\"が含まれる書籍の数: {wBooks.Count(x => x.Title.Contains("C#"))}冊");
```
[追加問題]
"C#" が大文字小文字、全角半角にかかわらずカウントされるように修正してみてください。

kurosu-being replied:
ありがとうございます！
追加問題を仮実装しました。
文字の大文字、小文字、半角全角は区別せずに判別できたのですが、＃の記号を半角全角を区別せずにカウントすることが出来ませんでした。
他に検討した方法は正規表現を用いる方法とMStrConvメソッド(Microsoft.VisualBasic名前空間のStringsクラス)などです。
メソッドで具体的に全角の♯を見つけて、半角の#に置き換えることは出来そうですが、他の記号でもに同様に書けるような方法が分かりませんでした。
アドバイス、ご指導をいただきたく存じます🙇‍♀️

kurosu-being replied:
※11/16追記
お疲れ様です。
こちらの質問ですが、正規表現の方法で＃の判定の区別が出来ない問題を解消しました。
ご確認よろしくお願いします。
（問題の原因は実装方法ではなくハッシュとシャープを混同していたことでした）

k-miyahara-being replied:
ちょっと無茶ぶりだったかもしれませんが、実現できたようで良かったです！

k-miyahara-being commented on 章末問題6/Practice6-2-3/Program6-2-3.cs:
```diff
@@ -0,0 +1,27 @@
+                Console.WriteLine("コレクションの中身がnullです。");
+                return;
+            }
+            Console.WriteLine($"タイトルに「C#」が含まれる書籍の平均ページ数: {wBooks.Where(x => x.Title.Contains("C#")).Average(x => x.Pages)}P");
```
以下のとき発生する例外に対応してください。
・一致する書籍がリストに存在しないとき
・Titleがnullのとき

kurosu-being replied:
ご指摘ありがとうございます。
ご指摘頂いた2つの例外に対応するように修正いたしました。

k-miyahara-being commented on 章末問題6/Practice6-2-1/Program6-2-1.cs:
```diff
@@ -0,0 +1,43 @@
+                new Book ( "一人で学ぶ並列処理プログラミング", 4800, 464 ),
+                new Book ( "フレーズで覚えるC#入門", 5300, 604 ),
+                new Book ( "私でも分かったASP.NET MVC", 3200, 453 ),
+                new Book ( "楽しいC#プログラミング教室", 2540, 348 ),
```
表記を統一してください。

・`Book`と`(` の間の空白は不要です。
・`()`内の両端の空白は不要です。
・`,`の後は空白を追加してください。

kurosu-being replied:
ご指摘ありがとうございます。
表記を統一いたしました。

k-miyahara-being commented on 章末問題6/Practice6-2-4/Program6-2-4.cs:
```diff
@@ -0,0 +1,24 @@
+                new Book ( "楽しいC#プログラミング教室", 2540, 348 ),
+            };
+
+            string wTargetBookTitle = wBooks.FirstOrDefault(x => x.Price >= 4000) != null ? $"価格が4000円以上の本で最初に見つかった書籍のタイトル: {wBooks.FirstOrDefault(x => x.Price >= 4000).Title}" : "4000円以上の書籍はありません。";
```
`wBooks.FirstOrDefault(x => x.Price >= 4000)`が2か所で書かれており、冗長になっています。
同じ処理が走ることになるので、その分2倍のコストがかかってしまいます。
こういう時は無理に三項演算子は使わず、一度変数に入れて扱うようにするとよいです。

kurosu-being replied:
ご指摘ありがとうございます。
処理をwMatchBooksに格納して修正いたしました。

k-miyahara-being commented on 章末問題6/Practice6-2-5/Program6-2-5.cs:
```diff
@@ -0,0 +1,26 @@
+                new Book ( "楽しいC#プログラミング教室", 2540, 348 ),
+            };
+
+            int wMaxPages = wBooks.Where(x => x.Price < 4000).Select(x => x.Pages).DefaultIfEmpty(0).Max();
+            string wTargetPageMessage = wMaxPages > 0 ? $"価格が4000円未満の本の中で最大ページ数: {wMaxPages}P" : "価格が4000円未満の本はありません。";
```
該当する要素なしのパターンもチェックされてますね👍

k-miyahara-being commented on 章末問題6/Practice6-2-6/Program6-2-6.cs:
```diff
@@ -0,0 +1,28 @@
+                new Book ( "楽しいC#プログラミング教室", 2540, 348 ),
+            };
+
+            var wTargetBooks = wBooks?.Where(x => x != null && x.Pages >= 400).OrderByDescending(x => x.Price).ToList();
```
nullチェックバッチリですね！

ここのToList()も不要と思います。

kurosu-being replied:
ご指摘ありがとうございます！
ToListを削除しました。

k-miyahara-being commented on 章末問題6/Practice6-2-6/Program6-2-6.cs:
```diff
@@ -0,0 +1,28 @@
+                ? "ページ数が400ページ以上の書籍を、価格が高い順に表示\n" + string.Join("\n", wTargetBooks.Select(x => $"書籍: {x.Title} 価格:{x.Price.ToString("#,0")}円"))
+                : "ページ数が400ページ以上の書籍は存在しません。";
+
+            Console.WriteLine(wResultMessage ?? "コレクションの中身がnullです。");
```
wResultMessageがnullにならないので、`?? "コレクションの中身がnullです。"` に到達することがありません。
wBooksがnullかどうかはもっと早い段階で判定できるので、早期returnしてしまうとよいです。

kurosu-being replied:
ご指摘ありがとうございます。
早期にwBooksの中身を判定して返すように修正いたしました。
Practice6-2-7も同様に変更しました。

k-miyahara-being reviewed (changes requested):
問2も見ました。
確認をお願いします。

kurosu-being committed:
#458736 [C#テキスト]6章演習問題

Practice6-2全問において、Listフォーマットを整えました。

kurosu-being committed:
#458736 [C#テキスト]6章演習問題

Practice6-2-2において追加問題を仮実装しました。

kurosu-being committed:
#458736 [C#テキスト]6章演習問題

Practice6-2-3において、一致する書籍がリストに存在しないときとTitleがnullのときに対応するように実装しました。
Practice6-2-4において、変数wMatchBooksの中に、wBooks.FirstOrDefault(x => x.Price >= 4000)の処理を入れました。
Practice6-2-6において、ToListを削除し、wBooksの中身を早期に判定するように修正しました。
Practice6-2-7において、wBooksの中身を早期に判定するように修正しました。

ko-kato-being commented on 章末問題6/Practice6-1-3/Program6-1-3.cs:
```diff
@@ -0,0 +1,15 @@
+        static void Main(string[] args) {
+            var wNumbers = new int[] { 5, 10, 17, 9, 3, 21, 10, 40, 21, 3, 35 };
+
+            foreach (string wStringNumber in wNumbers.Select(x => x.ToString())) {
+                Console.WriteLine(wStringNumber);
+            }
```
問題の意図には沿わないのかもしれないですけど、実状としては↓のように書くことの方が多いと思います。
```
foreach (int wNumber in wNumbers) {
    Console.WriteLine(wNumber);
}
```
Console.WriteLine() にはたくさんオーバーロードがあって、引数が string 型でも int 型でも同じように出力できます。
int 型でも出力できるのは、出力時に Int32.ToString() を呼び出して string 型に変換してくれているからですね。
https://learn.microsoft.com/en-us/dotnet/api/system.console.writeline?view=netframework-4.7.2#system-console-writeline(system-int32)

kurosu-being replied:
ご指摘ありがとうございます！
ご提案頂いたように修正いたしました。数字がintに格納されている方が違和感なくコードが読めますね。
また、ずっとConsole.WriteLine() を使っていたのに一度も調べたことがないことに気が付きました。使っているメソッドのオーバーロードを見る癖をつけたいです。

ko-kato-being commented on 章末問題6/Practice6-2-3/Program6-2-3.cs:
```diff
@@ -0,0 +1,31 @@
+                return;
+            }
+
+            var wMatchingBook = wBooks.FirstOrDefault(x => x.Title?.Contains("C#") == true);
+            string wMessage = wMatchingBook != null ? $"タイトルに「C#」が含まれる書籍の平均ページ数: {wBooks.Average(x => x.Pages)}P" : "一致する書籍が存在しません。";
+
+            Console.WriteLine(wMessage);
```
Title の null チェックは不要かな。
あと、== ture も書かない方がかっこいいかも:blush:

参考までに、別の書き方一例です。
※修正は不要です。
```
int? wAveragePages =  wBook.Where(x => x.Title.Contains("C#")).DefaultIfEmpty()?.Average(x => x.Pages);
if (wAveragePages == null) {
    Console.WriteLine("一致する書籍が存在しません。");
} else {
    Console.WriteLine($"タイトルに「C#」が含まれる書籍の平均ページ数: {wAveragePages}P");
}
```

(2023/11/20 追記)
DefaultIfEmpty()が要素を返すと勘違いして書いたコードですので、期待通りに動作しないと思います。

kurosu-being replied:
ご指摘ありがとうございます。
Titleのnullチェックを削除し修正しました。※11/17 再修正

コード例もありがとうございます！
特にint?の書き方が初見だったので驚きました。Nullable<T>型について適切に扱えるよう勉強したいと思います。

kurosu-being replied:
※11/17修正　
Titleのnullチェックですが、ContainsメソッドのところでTitleがnullの書籍があった場合に例外が発生したので
nullチェックを追加して再度修正いたしました。
お手数をおかけしますが、再度ご確認いただけると幸いです🙇‍♀️

yuki-tanaka-being replied:
@ko-kato-being 
`DefaultIfEmpty` のあとの `?` は不要ですね。
要素ではなく、列挙が返ってくるので `wBook` が `null` 以外でここの LINQ のメソッドチェーンが `null` になることはないかな。
逆に、`Where` で要素が空なら、`DefaultIfEmpty` で `null` が1個の列挙になります。
なので、`Average` で `NullRefarenceException` が発生します。
`DefaultIfEmpty` を使うなら `Select` で `Pages` だけにしてから `Average` に通すか、引数に `Pages` が `0` とか `-1` の `Book` を `new` して指定するとかがあります。
```
int wAveragePages =  wBook.Where(x => x.Title.Contains("C#")).DefaultIfEmpty(new Book("", 0, -1)).Average(x => x.Pages);
if (wAveragePages == -1) {
    Console.WriteLine("一致する書籍が存在しません。");
} else {
    Console.WriteLine($"タイトルに「C#」が含まれる書籍の平均ページ数: {wAveragePages}P");
}
```
ただし、個人的にはあまりおすすめしません。可読性が下がるためです。
この方法だと `-1` という数字に該当する本が無いという情報を与えることになります。
`Where` の結果 Book が1つもないと表現した方が、実態をコードでそのまま表現することができ、読んでいて何をしたいのか一目瞭然になるからです。
`-1` が何を示しているのかというのを読み解く必要がなくなります。

ko-kato-being replied:
@yuki-tanaka-being 
ご指摘ありがとうございます。
[リファレンス](https://learn.microsoft.com/ja-jp/dotnet/api/system.linq.enumerable.defaultifempty?view=netframework-4.7.2)を読み間違えてしまいました。
(DefaultIfEmpty()をあまり使ったこと無いのが露呈してしまいました:sweat:)

@kurosu-being 
間違ったコード書いてしまってすみませんね。
冒頭のコード例は反面教師としてください:pray:

> Titleのnullチェックですが、ContainsメソッドのところでTitleがnullの書籍があった場合に例外が発生したので
> nullチェックを追加して再度修正いたしました。

null が来たらどうなるかという観点で、Title が null の Book を追加してくれたんですかね。
目の付け所がすごく良いと思います:blush:

ただ、私としてはやはり null チェックをしない方が良いと思っています。
なぜなら、Book クラスを生成する際に、Title に null を渡すことは無いと思うからです。
言い換えると、タイトルの無い本はあり得ないということです。

タイトルの無い本の存在を認めてしまうと、Book クラス の Title を参照する際に、毎回 null チェックが必要になってしまいます。
本来あり得ないケースなのに、余計な手間が増えてしまいます:sweat_smile:

とはいえ、Title に null が入っていると、書いてくれた通りどこかで NullReferenceException が発生する可能性がありますよね。
こういう場合は、Book クラスのインスタンスを生成する時点で、Title が null だったら例外をスローしてしまうのも一つの手です。
こうすることで、Book クラスのインスタンス生成方法がおかしいことに気づけます。
```
public Book(string vTitle, int vPrice, int vPages) {
    this.Title = vTitle ?? throw new ArgumentNullException(nameof(vTitle), "Title cannot be null.");
    this.Price = vPrice;
    this.PageNumber = vPages;
}
```

kurosu-being replied:
@yuki-tanaka-being  @ko-kato-being @k-miyahara-being 
ご指摘、ご解説ありがとうございます！
加藤さんにご提案頂いたコードを使用させて頂き、
BookクラスでTitleがnullの時に例外をスローし、MainでTitleのnullチェックを逐一行わない方法で変更いたしました。

nullチェックをすることと同じぐらい、nullチェックをするためにコードを見づらくしていないかの吟味も重要なのですね。
また、局所的なnullチェックで良いのかといった、コードを再利用しようとしたときにどうなるか、という観点を考えるのが難しく感じるので今回の状況を覚えて理解を深めていきたいと思います。

ko-kato-being commented on 章末問題6/Practice6-2-4/Program6-2-4.cs:
```diff
@@ -0,0 +1,25 @@
+                new Book("楽しいC#プログラミング教室", 2540, 348),
+            };
+
+            var wMatchBooks = wBooks.FirstOrDefault(x => x.Price >= 4000);
+            string wTargetBookTitle = wMatchBooks != null ? $"価格が4000円以上の本で最初に見つかった書籍のタイトル: {wMatchBooks.Title}" : "4000円以上の書籍はありません。";
+            Console.WriteLine(wTargetBookTitle);
```
wTargetBookTitle という変数が、中身と合っていないような気がしますね:sweat_smile:
実際には、タイトルではなくメッセージですね。


こっちも別の書き方一例です。
※修正不要です。

```
string wTitle =  wBooks.FirstOrDefault(x => x.Price >= 4000)?.Title;
if (string.IsNullOrEmpty(wTitle)) {
    Console.WriteLine("4000円以上の書籍はありません。");
} else {
    Console.WriteLine($"価格が4000円以上の本で最初に見つかった書籍のタイトル: {wTitle}");
}
```

これくらいスコープが小さければ、簡単な変数名で良いかもしれないですね:bulb:

kurosu-being replied:
ご指摘ありがとうございます。
wTargetBookTitle →wMessageに変更いたしました。
書き方の例もありがとうございます！Nullの時に返すものがあるメソッドを使うと確実に例外を排除出来ますね。勉強になります！

ko-kato-being commented on 章末問題6/Practice6-2-1/Book.cs:
```diff
@@ -0,0 +1,30 @@
+        /// 価格
+        /// </summary>
+        public int Price { get; private set; }
+        /// <summary>
+        /// ページ数
+        /// </summary>
+        public int Pages { get; }
```
「ページ数」であれば、PageNumber や PageCount の方が適切な気がしますね。
接尾辞の「s」は、基本的にコレクション型の変数名に使用します。
型がリストや配列でない時は、「s」を付けない方が良いです。
誤解される可能性がありますので:sweat_smile:

kurosu-being replied:
ご連絡ありがとうございます。
Pages→PageNumber に変更いたしました。

> 接尾辞の「s」は、基本的にコレクション型の変数名に使用します。

他者が見た時に誤解を招かないようにするためにも、命名のセオリーを知っていることは重要ですね。
覚えます🦾

ko-kato-being commented on 章末問題6/Practice6-2-1/Book.cs:
Book クラスがいくつかあるんですが、すべて同じですか？
同じであれば、各プロジェクトで使いまわす方が良いかもですね。
いわゆるDRY原則です:bulb:

kurosu-being replied:
すべて同じものであった為各プロジェクトで使いまわすように修正しました。
クラスの重複は新しい問題に取り組むときに気になっていたことだったので、解決方法がわかってスッキリしました😀

kurosu-being committed:
#458736 [C#テキスト]6章演習問題

Practice6-2-2において不要なusingを削除しました。

kurosu-being committed:
#458736 [C#テキスト]6章演習問題

Practice6-2に以降においてBookのプロパティPages→PageNumberに変更いたしました。
Practice6-1-3において、変数の命名を変更し、foreachの中の変数の型をintに変更しました。
Practice6-2-3/Program6-2-3.csにおいて、wTargetBookTitle →wMessageに変更しました。

kurosu-being committed:
#458736 [C#テキスト]6章演習問題

Practice6-2-2に以降において参照を追加し、Practice6-2-1のBookクラスを利用できるように修正しました。

kurosu-being committed:
#458736 [C#テキスト]6章演習問題

Practice6-2-2ににおいて、「＃」についてハッシュではなくシャープを検索する対応しました。

kurosu-being committed:
#458736 [C#テキスト]6章演習問題

Practice6-2以降ににおいて、Listのフォーマットをそろえました。

kurosu-being committed:
#458736 [C#テキスト]6章演習問題

Practice6-2-3ににおいて、Titleのnullチェックを追加しました。

kurosu-being committed:
#458736 [C#テキスト]6章演習問題

Practice6-2-3ににおいて、Titleのnullチェックの書き方を修正しました。

yuki-tanaka-being commented on 章末問題6/Practice6-2-3/Program6-2-3.cs:
```diff
@@ -0,0 +1,32 @@
+            }
+
+            var wMatchingBook = wBooks.Where(x => x.Title != null && x.Title.Contains("C#"));;
+            string wMessage = wMatchingBook.Any() ? $"タイトルに「C#」が含まれる書籍の平均ページ数: {wMatchingBook.Average(x => x.PageNumber)}P" : "一致する書籍が存在しません。";
```
`wMatchingBook` が2回評価されています。
`Any()` と `Average()` の実行時です。
LINQでの評価は基本的に1回になるようにしたほうがよいです。
理由は2点あります。
・評価の度にコストがかかる。 ※今回は Any は実質O(1)なので問題ないですが。
・評価の度に結果が変わる恐れがあります。　※遅延評価のため。
今回は、どちらの影響も受けないので今のままでも問題はありませんが、意識して対応できているならGood！
そうでなければ、評価するタイミングとそれに付随して考慮すべきことは必須で押さえておいた方がよいです。
LINQはよく使うので、遅延評価の理解度は早めに上げておくのがおすすめです。

対処方法は先に1回評価した結果を保持して、それを利用するです。
一般的には `ToList()` した結果を使うことが多いです。

kurosu-being replied:
ご指摘ありがとうございます！
wMatchingBookをリスト化し、評価後の結果を用いてメソッドをを使うように変更いたしました。
（今回はどちらの影響も受けないので問題ないとのことでしたが、意識して行ったものではなかった為学習もかねてList化しました）
遅延実行のメソッドを多く扱うLINQは特に、何度も評価してコストを増していないか気にする必要があるのですね。
また、皆さんのご指摘を受け、どこでnullが発生しうるかやnullが入った時の戻り値をまだ把握しきれていない部分があると感じたので、LINQの理解度をもっと上げたいと思います。

yuki-tanaka-being commented on 章末問題6/Practice6-2-6/Program6-2-6.cs:
```diff
@@ -0,0 +1,34 @@
+                return;
+            }
+
+            var wTargetBooks = wBooks?.Where(x => x.PageNumber >= 400).OrderByDescending(x => x.Price);
+
+            string wResultMessage = wTargetBooks?.Any() == true
+                ? "ページ数が400ページ以上の書籍を、価格が高い順に表示\n" + string.Join("\n", wTargetBooks.Select(x => $"書籍: {x.Title} 価格:{x.Price.ToString("#,0")}円"))
+                : "ページ数が400ページ以上の書籍は存在しません。";
```
2回評価されているという 6.2.3 と同じ内容ですが、こちらはコストがかかってしまっているパターンです。
LINQ で列挙対象に対する処理がいつ行われるかといった細かい話になりますが、Where や Select といった逐次処理可能なものは1個ずつ処理が流れていきます。
Any や First みたいなものは条件と一致するものが最初に見つかった時点まで一気に処理されます。
OrderBy や Max などは、すべての要素を一気に処理します。
そのため、今回だと `wTargetBooks?.Any()` の時点で全要素に対して処理がされます。
その後に、`wTargetBooks.Select(x => $"書...)` で、再び全要素に対して処理がされます。
O(2n) になっています。

どの LINQ がどういう性質かは、そのメソッドが何をするものかをよく考えるとだいたい分かると思います。
例えば、
Where や Select は各要素間の依存はなく、要素1個1個に対する処理なので、1個ずつ次の処理に流すことができます。
OrderBy や Max は、全要素を確認しないと最終的な結果が確定できないので、一気に処理が走ります。
といった具合になります。

kurosu-being replied:
ご指摘ありがとうございます！
wMatchingBookをリスト化し、評価後の結果を用いてAny()メソッドをを使うように変更いたしました。
全要素を確認しないと出来ない処理は特に注意して、LINQ についてもっと理解を深めたいと思います。

yuki-tanaka-being commented on 章末問題6/Practice6-2-6/Program6-2-6.cs:
```diff
@@ -0,0 +1,34 @@
+
+            var wTargetBooks = wBooks?.Where(x => x.PageNumber >= 400).OrderByDescending(x => x.Price);
+
+            string wResultMessage = wTargetBooks?.Any() == true
```
特に理由がないなら `== true` は不要かな？

kurosu-being replied:
ご指摘ありがとうございます！
== trueを削除して修正しました。
どうしても?.を使いたい脳になっていました。。。🙃

yuki-tanaka-being commented on 章末問題6/Practice6-2-2/Program6-2-2.cs:
```diff
@@ -0,0 +1,43 @@
+    //追加問題 "C#" が大文字小文字、全角半角にかかわらずカウントされるように修正してみてください。
+    class Program {
+        static void Main(string[] args) {
+            var wBooks = new List<Book>{
+                //確認の為C#の表記を大文字小文字半角全角を試しています
+                new Book("C#プログラミングの常識", 3800, 378),
+                new Book("ラムダ式とLINQの極意", 2500, 312),
+                new Book("ワンダフル・c#ライフ", 2900, 385),
+                new Book("一人で学ぶ並列処理プログラミング", 4800, 464),
+                new Book("フレーズで覚えるｃ#入門", 5300, 604),
+                new Book("私でも分かったASP.NET MVC", 3200, 453),
+                new Book("楽しいＣ♯プログラミング教室", 2540, 348),
+            };
```
コレクションの初期化子を使ってくれていますね👍 
配列限定ですが、初期化子があると型推論できるので↓みたいに書くこともできます。
```
var wBooks = new [] {
    new Book("C#プログラミングの常識", 3800, 378),
          :
    new Book("楽しいＣ♯プログラミング教室", 2540, 348),
}
```
こういうテストデータ的なコレクションを用意するときには個人的によく使います。

kurosu-being replied:
こんな書き方もあるのですね😲
書き方の幅が広がります、ありがとうございます！

yuki-tanaka-being reviewed (commented):
だいぶ出遅れましたが、気になった点をコメントしました。

kurosu-being committed:
#458736 [C#テキスト]6章演習問題

Practice6-2-3において、26行目のwMatchingBookをリスト化し、評価後の結果を用いてAny()メソッドをを使うように変更いたしました。
Practice6-2-6において、wTargetBookをリスト化しました。
Practice6-2-6において、wSelectMessageであらかじめ文字列を作成する記述をコメントにしました。（学習の為のコメントで、後に削除します。）

kurosu-being committed:
#458736 [C#テキスト]6章演習問題

Practice6-2-6において、不要なコメントを削除しました。

kurosu-being committed:
#458736 [C#テキスト]6章演習問題

Practice6-2-6において、LINQの学習の為のコメントを一時的に記述しました。（後で削除してください）

kurosu-being committed:
#458736 [C#テキスト]6章演習問題

Practice6-2-6において、不要なコメントを削除しました。

ko-kato-being commented on 章末問題6/Practice6-1-1/Program6-1-1.cs:
```diff
@@ -0,0 +1,19 @@
+
+            //追加問題　要素があるかどうかを判定する処理を追加
+            string wJudgeNumbers = wNumbers.Any() ? wNumbers.Max().ToString() : "要素がありません";
+            Console.WriteLine(wJudgeNumbers);
```
ここも ToString() しない方が見た目はすっきりしますね。
まあこれは好みの範疇なので修正必須ではないです！

kurosu-being replied:
ご指摘ありがとうございます！
色々な書き方に慣れたいので修正してみました。
確認して頂けると幸いです🙇‍♀️

ko-kato-being commented on 章末問題6/Practice6-1-2/Program6-1-2.cs:
```diff
@@ -0,0 +1,26 @@
+    //Practice6-1-2 最後から2つの要素を取り出して表示してください。
+    class Program {
+        static void Main(string[] args) {
+            var wNumbers = new int[] { 5, 10, 17, 9, 3, 21, 10, 40, 21, 3, 35 };
+
+            if (wNumbers == null) {
+                Console.WriteLine("コレクションの中身がnullです");
+                return;
+            }
```
null チェックをとても意識してくれているのがひしひしと伝わってきます！
それはすごく良いことなのですが、ここでは null チェックしなくて良いと思います。

すぐ上のコードで、wNumbers を初期化しており、null になることはあり得ないからです。
ここで null チェックをしてしまうと、今後の演習でコレクションクラスを生成する度に、すべて null チェックしないといけなくなっちゃいます:scream:

kurosu-being replied:
ご指摘ありがとうございます！
nullチェックを削除し修正しました。
null チェックを意識したものの、発生し得ないときにチェックしてしまうと、逆にコストがかかっているコードになってしまいますね。

また、wNumbersを初期化する際に無理やりnullを入れられると思っていたのですが、実際試してみるとコンパイルエラーが出て入れられませんでした。
（int型にはそもそもnullが代入できませんね😇）

ko-kato-being commented on 章末問題6/Practice6-1-4/Program6-1-4.cs:
```diff
@@ -0,0 +1,20 @@
+        static void Main(string[] args) {
+            var wNumbers = new int[] { 5, 10, 17, 9, 3, 21, 10, 40, 21, 3, 35 };
+
+            if (wNumbers == null) {
+                Console.WriteLine("コレクションの中身がnullです");
+                return;
+            }
```
ここも同じく null チェック無くて良いと思います。
と書いたんですが、↓の指摘によって練習で null チェックしてたんですね。
https://github.com/kurosu-being/TextC/pull/7/files?file-filters%5B%5D=.cs&show-viewed-files=true#r1387485936

ということであれば、いまのままにしておいてもらってOKです。
ただ、宮原くんも書いてくれていますが、本来は null チェックは不要だということは改めて認識してもらうと良いかな。

kurosu-being replied:
ご指摘ありがとうございます！
ご提案頂いた通り、この箇所に関してはこのままにしておこうと思います。
nullチェックが必要な場合とそうでない場合について、もっと理解を深めたいと思います。

ko-kato-being commented on 章末問題6/Practice6-1-4/Program6-1-4.cs:
```diff
@@ -0,0 +1,20 @@
+                return;
+            }
+            Console.WriteLine("小さい順に並べた最初の3つの要素:");
+            foreach (int wSortThreeNumber in wNumbers.OrderBy(x => x).Take(3)) {
```
細かい話ですが、変数名は wNumber で十分かなと思いました。
コレクションループの中では、その要素が「小さい順に並べた最初の3つの要素」なのかどうかは判別できないからです。
(その情報はループの中では必要ないはず。)

もう一つ別の観点から、
仮に「4つの要素」に変更したとすると、変数名まで変更する必要があります。

命名ってはじめはホントに難しいと思うけど、じょじょに慣れていってください:relieved:

kurosu-being replied:
ご指摘ありがとうございます！
wNumberに変更いたしました。
命名について初期よりは分かってきたつもりだったのですが、未だに迷走しているので皆さんのご指摘やテキストで慣れていきたいと思います。

ko-kato-being commented on 章末問題6/Practice6-2-6/Program6-2-6.cs:
```diff
@@ -0,0 +1,39 @@
+                new Book("楽しいC#プログラミング教室", 2540, 348),
+            };
+
+            if (wBooks == null) {
+                Console.WriteLine("コレクションの中身がnullです。");
+                return;
+            }
+
+            var wTargetBooks = wBooks?.Where(x => x.PageNumber >= 400).OrderByDescending(x => x.Price).ToList();
```
`wBook?` の null チェックは不要ですね。
すぐ上で既に null チェックを行っていますので。

kurosu-being replied:
ご指摘ありがとうございます！
nullチェックを削除しました。

ko-kato-being commented on 章末問題6/Practice6-2-7/Program6-2-7.cs:
```diff
@@ -0,0 +1,34 @@
+                new Book("楽しいC#プログラミング教室", 2540, 348),
+            };
+
+            if (wBooks == null) {
+                Console.WriteLine("コレクションの中身がnullです。");
+                return;
+            }
+
+            var wTargetBooks = wBooks?.Where(x =>  x.Title.Contains("C#") && x.PageNumber <= 500);
```
ここも同じく `wBooks?` の null チェックは不要ですね。

kurosu-being replied:
ご指摘ありがとうございます！
同様に、null チェックを削除いたしました。

ko-kato-being commented on 章末問題6/Practice6-2-7/Program6-2-7.cs:
```diff
@@ -0,0 +1,34 @@
+            }
+
+            var wTargetBooks = wBooks?.Where(x =>  x.Title.Contains("C#") && x.PageNumber <= 500);
+            string wResultMessage = wTargetBooks.Any() == true
```
` == true` が残っているのは修正漏れかな？

kurosu-being replied:
修正漏れです、失礼しました🙇‍♀️
削除して修正いたしました。

ko-kato-being reviewed (changes requested):
改めてソース確認してみました。
またコメントが増えてしまって申し訳ないのですが、確認お願いします。

コメントが増えてきてしまったので、終わった会話は Resolve conversation してもらうと良いかなと思います。
本来ならレビュー者が実施するのが良いと思うのですが、ボタンが表示されておりませんので。
よくわからなかったら聞いてください！

kurosu-being committed:
#458736 [C#テキスト]6章演習問題

Practice6-1-1において、ToString()を削除して書き直しました。
Practice6-1-2において、nullチェックを削除しました。
Practice6-1-4において、wSortThreeNumber→wNumberに修正しました。
Practice6-2-6において、wBook?.のnullチェックを削除しました。
Practice6-2-7において、==trueを削除しました。

kurosu-being committed:
#458736 [C#テキスト]6章演習問題

Practice6-2-2において、フォーマットを修正しました。

kurosu-being committed:
#458736 [C#テキスト]6章演習問題

Practice6-2-1において、BookクラスをTitleにnullが代入された時に対応するよう修正しました。
Practice6-2-3において、Titleのnullチェックを削除しました。

yuki-tanaka-being commented on 章末問題6/Practice6-2-6/Program6-2-6.cs:
```diff
@@ -0,0 +1,39 @@
+                return;
+            }
+
+            var wTargetBooks = wBooks.Where(x => x.PageNumber >= 400).OrderByDescending(x => x.Price).ToList();
+            var wSelectMessages = wTargetBooks.Select(x => $"書籍: {x.Title} 価格:{x.Price.ToString("#,0")}円");
+         
+            if (!wTargetBooks.Any()) {
+                Console.WriteLine("ページ数が400ページ以上の書籍は存在しません。");
+                return;
+            }
+
```
26行目と28-31行目の順番を入れ替えた方がよいです。
`wSelectMessages` が必要になるのは 34行目であり、それまでは不要だからです。
変数宣言の基本的な考え方として、初めて必要になった時に用意する（スコープが最も小さくなるようにする）と覚えておくとよいです。
理由は、スコープを小さくすることでバグを埋め込みにくくなるのと、可読性が上がるからです。
頭の中でコードを実行する際に、余計なスタックを1個多く積む必要があります。
人間が短期記憶できる個数は限られるので、できるだけスタックを積まなくてもよい方が可読性が高いです。

今回の場合は、34行目に記載するか、変数に取らずに直接 foreach に渡すがよいです。
※元の実装の名残でしょうが、fereach で回すなら Select も不要ですね。
　WriteLine 時に文字列を構築するのが一般的です。

kurosu-being replied:
（修正途中でメッセージを送ってしまいました。失礼しました。）
再度修正を行ったものを反映させました。ご確認下さい🙇‍♀️

> 変数宣言の基本的な考え方として、初めて必要になった時に用意する（スコープが最も小さくなるようにする）と覚えておくとよいです。

改めてコードを見返すと確かに違和感があります。宣言の位置と実際に使用する位置が離れていたからだったのですね。基本の考えもってコードが書けるよう定着させたいと思います。

kurosu-being committed:
#458736 [C#テキスト]6章演習問題

Practice6-2-2において、wBooksがnullの時のメッセージを変更しました。
Practice6-2-3において、wBooksがnullの時のメッセージを変更しました。
Practice6-2-6において、wSelectMessagesを削除しました。
Practice6-2-7において、wBooksがnullの時のメッセージを変更しました。

kurosu-being committed:
#458736 [C#テキスト]6章演習問題

Practice6-2-7において、ToList()を追加しました。
